_source/_posts/2021-11-08-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,562 @@
+
+Next, you will need to generate the `api-token` value. Log in to your Okta administrator console as a superuser and select **Security** > **API** > **Tokens (Tab)** from the navigation menu. Next, click the **Create Token** button, give your token a name, click **Create Token**, and copy the newly generated token. Save this in a separate `.tfvars` file excluded from Git or in an environment variable named `TF_VAR_api_token`.
+
+The Okta provider is now configured and ready to go.
+
+### Create groups
+
+Now we need some groups to differentiate and map different kinds of users who want to access our clusters. Let's say we have a group of administrators with full access to the cluster and another group of users who have limited access. You can have any number of groups as per your needs. The configuration below will create two groups. The group's privileges will be defined using Kubernetes RBAC policies on the cluster, which we will do later.
+
+```hcl
+# Set up OKTA groups
+resource "okta_group" "k8s_admin" {
+  name        = "k8s-admins"
+  description = "Users who can access k8s cluster as admins"
+}
+
+resource "okta_group" "k8s_restricted_users" {
+  name        = "k8s-restricted-users"
+  description = "Users who can only view pods and services in default namespace"
+}
+```
+
+### Assign users to the groups
+
+The below snippet looks up existing users and adds them to the groups. You can add any number of users at this stage, or you can skip adding users and do it via the Okta Admin Console later. For this exercise, I'm fetching existing users. You could also create new users using the `okta_user` resource.
+
+```hcl
+# Assign users to the groups
+data "okta_user" "admin" {
+  search {
+    name  = "profile.email"
+    value = "<your_admin_user_email>"
+  }
+}
+
+resource "okta_group_memberships" "admin_user" {
+  group_id = okta_group.k8s_admin.id
+  users = [
+    data.okta_user.admin.id
+  ]
+}
+
+data "okta_user" "restricted_user" {
+  search {
+    name  = "profile.email"
+    value = "<another_user_email>"
+  }
+}
+
+resource "okta_group_memberships" "restricted_user" {
+  group_id = okta_group.k8s_restricted_users.id
+  users = [
+    data.okta_user.restricted_user.id
+  ]
+}
+```
+
+### Create an OIDC application
+
+Now that our groups are in place let's create an OIDC application. We will set the application type to `native` and use PKCE as client authentication, which is much more secure than using a client secret. We will also set the redirect URIs to `localhost:8000` so that we can work with kubectl locally. We should also assign the groups we created earlier to this application here. Finally, we can capture the client ID of the created app using an output variable.
```

Comment:
```suggestion
Now that our groups are in place, let's create an OIDC application. We will set the application type to `native` and use PKCE as client authentication, which is much more secure than using a client secret. We will also set the redirect URIs to `localhost:8000` so that we can work with kubectl locally. We should also assign the groups we created earlier to this application here. Finally, we can capture the client ID of the created app using an output variable.
```
comma